In [73]:
mit_syllabus_corpus = []
mit_syllabus_corpus_words = []

In [75]:
from bs4 import BeautifulSoup

for number in range(2494):

    HTMLFile = open("concat/" + str(number) + ".html", "r")
    index = HTMLFile.read()
    S = BeautifulSoup(index, 'lxml')

    if S.body != None:
        mit_syllabus_corpus.append(S.body.get_text().replace("\n", "").replace("\xa0", " "))
        mit_syllabus_corpus_words += S.body.get_text().replace("\n", "").replace("\xa0", " ").split()

In [56]:
# Code Adapted from COLX 521 Lecture 4

def average_word_length(words):
    total_words = 0
    total_chars = 0
    for word in words:
        total_words += 1
        total_chars += len(word)
    return total_chars/total_words

In [60]:
from nltk.corpus import treebank, webtext

In [113]:
print("webtext")
print("Number of words:")
len(webtext.words())

webtext
Number of words:


396733

In [114]:
print("treebank")
print("Number of words:")
len(treebank.words())

treebank
Number of words:


100676

In [115]:
print("MIT syllabus corpus")
print("Number of words:")
len(mit_syllabus_corpus_words)

MIT syllabus corpus
Number of words:


6256086

For number of words, MIT syllabus corpus has 6256086 words, treebank corpus has 100676 words, and webtext corpus has 396733 words. I think the reason why MIT syllabus corpus has much more words is because there are around 2500 syllabuses and each syllabus has very detailed description of a course.

In [120]:
print("treebank")
print("Average word length:")
average_word_length(treebank.words())

treebank
Average word length:


4.406154396281139

In [121]:
print("MIT syllabus corpus")
print("Average word length:")
average_word_length(mit_syllabus_corpus_words)

MIT syllabus corpus
Average word length:


6.143507298333175

In [122]:
print("webtext")
print("Average word length:")
average_word_length(webtext.words())

webtext
Average word length:


3.552701691061747

For average word length, MIT syllabus corpus's is 6.143507298333175, treebank's is 4.406154396281139, webtext's is 3.552701691061747. I think the reason why the average word length of MIT syllabus corpus is the highest is because it has more proper nouns including author's name and academic terminology. Webtext has the lowest may be because most of the words in webtext come from online discussion forums and they are mostly simple daily English words.

In [135]:
# Code adapted from COLX 521 Lecture 4

def get_simple_stats(corpus):
    #num_chars = sum([len(word) for word in corpus.words()])
    num_words = len(corpus.words())
    #num_sents = len(corpus.sents())
    num_texts = len(corpus.fileids())
    #print(num_texts)
    print("average text length")
    print(num_words/num_texts)

In [136]:
print("treebank")
get_simple_stats(treebank)

treebank
average text length
505.90954773869345


In [137]:
print("webtext")
get_simple_stats(webtext)

webtext
average text length
66122.16666666667


In [138]:
print("MIT syllabus corpus")
print("average text length")
print(len(mit_syllabus_corpus_words)/len(mit_syllabus_corpus))

MIT syllabus corpus
average text length
2560.821121571838


For average text length, MIT syllabus corpus's is 2560.821121571838, webtext's is 66122.16666666667, treebank's is 505.90954773869345. I think the reason why webtext's is the highest is because it has least amount of texts which is only 6.

In [131]:
# Code adapted from COLX 521 Lecture 4

def type_token_ratio(words):
    types = list(set(words))
    return len(types) / len(words)

In [132]:
print("MIT syllabus corpus")
print("type token radio")
type_token_ratio(mit_syllabus_corpus_words)

MIT syllabus corpus
type token radio


0.11294314048751887

In [133]:
print("treebank")
print("type token radio")
type_token_ratio(treebank.words())

treebank
type token radio


0.12324685128531129

In [134]:
print("webtext")
print("type token radio")
type_token_ratio(webtext.words())

webtext
type token radio


0.05428840051117502

For type token radio, MIT syllabus corpus's is 0.11294314048751887, webtext's is 0.05428840051117502, treebank's is 0.12324685128531129. MIT syllabus corpus's is close to treebank's and webtext's is much lower. I think MIT syllabus corpus and treebank have higher lexical richness may be because they have more proper nouns and terminologies.

In [139]:
# Code adapted from COLX 521 Lab 2 solution

import collections

def get_unigram_probs(words):
    counts = collections.Counter([word.lower() for word in words])
    total = sum(counts.values())
    return {word:count/total for word,count in counts.items()}

def combine_probs(prob1,prob2):
    all_words = set(prob1.keys())
    all_words.update(prob2.keys())
    return {word:(prob1.get(word,0) + prob2.get(word,0))/2 for word in all_words}

def subtract_probs(prob1, prob2):
    all_words = set(prob1.keys())
    all_words.update(prob2.keys())
    return {word:prob1.get(word,0) - prob2.get(word,0) for word in all_words}

probs_dicts = []
#for category in categories:
probs_dicts.append(get_unigram_probs(mit_syllabus_corpus_words))
probs_dicts.append(get_unigram_probs(treebank.words()))
probs_dicts.append(get_unigram_probs(webtext.words()))

categories = ["MIT syllabus corpus", "Treebank", "Webtext"]

for i in range(3):
    current_probs = probs_dicts[i]
    other_probs = probs_dicts[:i] + probs_dicts[i+1:]
    comb_probs = combine_probs(other_probs[0],other_probs[1])
    sub_probs = subtract_probs(current_probs, comb_probs)
    sub_sorted_words = sorted(sub_probs.keys(),key=lambda x: sub_probs[x])
    print(categories[i])
    print(sub_sorted_words[-100:])

MIT syllabus corpus
['chapters', 'urban', 'learning', 'their', 'd.', 'write', 'case', 'video', 'de', 'may', 'presentation', 'c.', 'data', 'american', 'assignments', 'sessions', 'development', 'used', 'exam', 'due', '5', 'study', '=', 'week,', 'reading', 'assignment', 'following', 'use', 'our', 'writing', 'questions', 'introduction', 'r.', 'google', 's.', 'part', 'discussion', 'ed.', '[preview', 'social', 'theory', 'session', 'p.', 'an', 'student', 'a.', 'readings', 'analysis', 'al.', 'research', 'm.', 'problem', 'edited', 'these', '(pdf)', '4', 'science', 'review', 'this', 'et', 'journal', 'york,', '(pdf', 'design', 'ny:', 'lecture', 'j.', 'mit', 'paper', 'we', '/', 'with', '3', 'as', 'project', 'or', 'final', 'each', '“the', 'are', 'how', 'university', 'set', 'by', 'chapter', 'be', 'class', 'students', 'press,', 'pp.', 'your', 'no.', 'course', 'for', 'will', 'isbn:', 'in', 'the', 'of', 'and']
Treebank
['many', 'bonds', 'but', 'investment', 'recent', 'companies', 'such', 'last', 'profi

According to the result above, I think the words associated with MIT syllabus corpus are mostly academic missions for the course, such as reading, assignment, discussion, lecture, etc. And I think those are the metadata of the MIT syllabus corpus. Because of those metadata, MIT syllabus corpus has more proper nouns and academic terminologies so that average word length and type token radio are higher. The words associated with treebank are from different backgrounds since they are from different journal articles. And the words associated with webtext are mostly daily English words and pronouns because they are collected mostly from online discussion forums.